### PR TITLE
Add option for install only

### PR DESCRIPTION
**Pull request by topherlandry**
### Description

Was having a problem such that the systemd handling of haproxy was not allowing for zero downtime haproxy reloads on long running TCP connections.

Decided to have the cookbook handle the installation and configuration, but did not want to have haproxy managed by a system service, so that I could fine tune the run / restart rules of haproxy on my own.

Added install_only boolean property to install recipe. This will only handle the installation and will not register haproxy as a service on the host system.

### Issues Resolved

N/A

### Contribution Check List

- [x] All tests pass.
- [x] New functionality includes testing.
- [x] New functionality has been documented in the README if applicable

---
This change is [Reviewable](https://reviewable.io/reviews/sous-chefs/haproxy/251)

### CODE REVIEW

**Comment by rshade** [comment]
what part of your init script are you changing to handle this?

**Comment by teknofire** [comment]
Would this work better if the template resource notified the service `reload` rather than do a `restart`?  Or are there other changes to the init script that you are having to make to support `0 downtime` restarts?

If we change the template to do a `reload` instead of `restart` then we need to make sure to handle the first run case where the service hasn't started yet.

**Comment by topherlandry** [comment]
To handle this, I've added additional start and reload recipes.

The start recipe just executes the following:
`sudo haproxy -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid`

The reload recipe executes the following:
`sudo haproxy -f /etc/haproxy/haproxy.cfg -p /run/haproxy.pid -sf $(cat /run/haproxy.pid)`

This keeps the current instance of haproxy running until any existing connections are complete, but starts an additional process to handle all new connections.

I believe I tested both the restart and reload options and both resulted in termination of existing connections. i could test this again though.

**Comment by rshade** [comment]
@topherlandry thank you, can you give us a little time to research. I want to hop on haproxy  irc and see what they are thinking. We are using their systemd shim, so it may have to be updated there. Are you using package or source btw?

**Comment by topherlandry** [comment]
I'm using source with the 1.7.8 release.

The zero downtime commands I'm using came from this article:
https://www.haproxy.com/blog/truly-seamless-reloads-with-haproxy-no-more-hacks/

**Comment by damacus** [comment]
Hey @topherlandry thanks for the contribution! 

Apart from the general "trying to figure out what we're trying to achieve here" with that code.

There are a couple of lint related issues noted. 

Thanks,
Dan

File: resources/install.rb [hunk 1]
```diff
@@ -79,7 +80,7 @@
     make_cmd << " USE_LINUX_TPROXY=#{use_linux_tproxy}"
     make_cmd << " USE_LINUX_SPLICE=#{use_linux_splice}"
 
-    extra_cmd = ' EXTRA=haproxy-systemd-wrapper' if node['init_package'] == 'systemd'
+    extra_cmd = ' EXTRA=haproxy-systemd-wrapper' if node['init_package'] == 'systemd' && !install_only
```
Comment (damacus) [inline resources/install.rb]: Where ever we access a resource from the current/new resource we really need to access it via `new_resource.X` that  way Chef knows which resource to use. IIRC this should actually be an error on current Chef 13.

File: resources/install.rb [hunk 2]
```diff
@@ -113,8 +114,10 @@
       mode '0644'
       source lazy { node.run_state['haproxy']['conf_template_source'][config_file] }
       cookbook lazy { node.run_state['haproxy']['conf_cookbook'][config_file] }
-      notifies :enable, 'poise_service[haproxy]', :immediately
-      notifies :restart, 'poise_service[haproxy]', :delayed
+      unless install_only
+        notifies :enable, 'poise_service[haproxy]', :immediately
```
Comment (damacus) [inline resources/install.rb]: So what's the trick here for only installing that? 

Apologies for not reading that post, any chance of a summary of what we're achieving here by *not* restarting. 

Hopefully a bunch of people learn from it asa  result :)

Comment (topherlandry) [inline resources/install.rb]: The problem that I encounter with the systemd shim is that I'm using haproxy in front of live streaming ingestion. The restart is terminating the connections. haproxy does have a method of restarting that does not terminate existing connections. This works for me, so I've made recipes for restarting and reloading.

That said, the recipes in this repo are very helpful for install / config generation and I'd like to continue using them for that.

File: resources/install.rb (hunk 1 above)
Comment (topherlandry) [inline resources/install.rb]: I'm sorry, I'm not sure what the change is being requested here. Is this in regards to `node['init_package']`? That's existing code, and as far as I can tell should still be supported in Chef 13, but I could be wrong on that and I'm happy to make that change.

That will also need to change on line 123 if that's the case, and I'd rather make a new PR for it.

If that's not the change requested then I really don't understand.

**Comment by damacus** [comment]
@topherlandry Thanks for the PR.
Quick update where we're at. 
Right now we ( @rshade  ) are looking at generally improving the systemd scripts. Hopefully this helps going forward.

**Comment by rshade** [comment]
from haproxy irc: 

> `
<rshade98> does anyone have a systemd service that does the -sf on reload example
<meineerde> rshade98: with systemd, you have to use haproxy-systemd-wrapper. With that, you can just send a SIGHUP to it to initiate the reload (which is the default for systemd)
<meineerde> rshade98: check the contrib/systemd directory of your haproxy source for an example systemd unit.
<rshade98> I have the wrapper, but what if I want to do -sf? 
<meineerde> rshade98: if you send a SUGHUP to the wrapper, it internally starts new processes and finishes the existing ones the same as -sf would.
<meineerde> Thus, with the wrapper, you don't need to handle this on your own at all (and you don't need manual pid files).
<rshade98> meineerde, thank you. Do you mind if I copy and paste your replies to this github issue: https://github.com/sous-chefs/haproxy/pull/251
<meineerde> rshade98: sure, although I'm not sure what these guys are attempting there :)
<meineerde> And I personally also use by own cookbook instead :) Although it needs some love (e.g. explicit systemd support *shudder*) it works just fine.
<meineerde> And since I use runit, I never had to deal with setting up haproxy as an actual systemd service...
<meineerde> https://github.com/meineerde-cookbooks/haproxy
`

**Pull request merged**